### ISSUE TITLE
REQ: Should `center` in `rowVars(..., rows=..., center=...) also be subsetted?

### ISSUE THREAD

**Issue by HenrikBengtsson** [Dongcan-Jiang/matrixStats#9]
# Background

A few functions take optional argument 'center', e.g. rowVars(), rowMads().  For example:

``` r
> x <- matrix(rnorm(20), nrow=5, ncol=4)
> mu0 <- rowVars(x)
[1] 0.6071190 0.1176126 0.6489145 0.7848270 0.5952322

> center <- rowMeans(x)
[1]  0.36222637  0.01655495  0.50372371 -0.47036082  0.13692486

> mu1 <- rowVars(x, center=center)
[1] 0.6071190 0.1176126 0.6489145 0.7848270 0.5952322

> stopifnot(identical(mu1, mu0))
```

If we choose to calculate the variance on a subset of _rows_ (not columns), we need to also subset `center`, e.g.

``` r
> rows <- 1:3
> mu1s <- rowVars(x[rows,], center=center[rows])
> mu1s
[1] 0.6071190 0.1176126 0.6489145
> stopifnot(mu1s == mu1[rows])
```
# What about `rowVars(x, rows=rows, center=center)`?

So, for the new subsetting features of matrixStats, should that be done automatically as well, e.g.

``` r
> rows <- 1:3
> mu1s <- rowVars(x, rows=rows, center=center)
```

?  It certainly looks neat to the user.  However, what if the user already got `center` calculated for those rows in `rows`, e.g.

``` r
> centerS <- rowMeans(x[rows,])
```

(the above is just an illustration for when the subset of centers are already available, so just imagine `centerS` is available).  Then such as user would need to expand `centerS` to `center` as follows if matrixStats takes care of the subsetting of `center` as well:

``` r
> rows <- 1:3
> center <- double(ncol(X))
> center[rows] <- centerS
> mu1s <- rowVars(x, rows=rows, center=center)
```
# Questions up for discussion

Q: Is the latter example rare enough to be ignored and is it better to automagically subset `center` as well?

Q: Can/should we support both approaches?  If user passes `length(center) == length(rows)` then it is assumed that subsetting is already done by the users.  If `length(center) == nrow(x)`, then subsetting should be done.  Would that be too confusing?  Could that lead to potentially hard to catch errors that we need to worry about, e.g. `rowMads(x, rows=nrow(x):1, center=center)`?!?

**Comment by HenrikBengtsson** [comment]
@Dongcan-Jiang and @hcorrada, please share your thoughts on this one as it's not clear to me what the user would assume and what the most common use case will be.

**Comment by Dongcan-Jiang** [comment]
@HenrikBengtsson @hcorrada Thanks. I indeed ignored that the center is a vector rather than a number. I agree that the `center` should also be subsetted in the way `w` is handled. I'm going to modify it.

**Comment by HenrikBengtsson** [comment]
I certainly lean the same way.  Let's do it - subset `center` as well.  If this turns out to be an issue later, we can always change.

Please close this issue when you've made the modifications.

PS. I think of `center` as somewhat different than weights `w`, because they apply in different directions, i.e. for row operations `center` is along rows (`length(center) == nrow(x)`) whereas weights along columns (`length(w) == ncol(x)`).

**Comment by Dongcan-Jiang** [comment]
This issue is fixed in 1259e65
